contrib/natsio/src/main/java/com/google/cloud/dataflow/contrib/natsio/NatsIO.java

```diff
@@ -0,0 +1,359 @@
+		}		
+
+		/**
+		 * Create a NatsSource instance which is bounded to total reading time to receive updates.
+		 * @param subject to subscribe and received updates
+		 * @param readTime bounds total reading time to receive updates from NATS server.
+		 * @param props contains NATS related properties. "servers" property is mandatory to access a remote NATS server.
+		 * @return a bounded NatsSource instance
+		 */
+		public static PTransform<PInput, PCollection<KV<String,String>>> withMaxReadtime(String subject, long readTime, Properties props) {
+			return from(subject, props).withMaxReadTime(Duration.standardSeconds(readTime));
+		}		
+		
+		public static class NatsSource extends UnboundedSource<KV<String,String>, UnboundedSource.CheckpointMark> {
+			private static final long serialVersionUID = 1L;
+			
+			private Properties props = null;
+			private Connection conn = null;
+			private String subject;
+			
+			public NatsSource(String subject, Properties props) {
+				this.subject = subject;
+				this.props = props;
+			}
+			
+			@Override
+			public NatsReader createReader(PipelineOptions options, UnboundedSource.CheckpointMark mark) {
+				return new NatsReader(subject);
+			}
+
+			@Override
+			public List<? extends UnboundedSource<KV<String,String>, UnboundedSource.CheckpointMark>> generateInitialSplits(
+					int numSplits, PipelineOptions options) throws Exception {
+				List<NatsSource> results = new ArrayList<NatsSource>();
+				// Create a NatsSource instance for each split.
+				int split = numSplits / 3;
+				for(int i = 0; i < split; i++)
+					results.add(new NatsSource(subject, props));
+				return results;
+			}
+
+			@Override
+			public Coder<UnboundedSource.CheckpointMark> getCheckpointMarkCoder() {
+				return null;
+			}
+
+			@Override
+			public Coder<KV<String,String>> getDefaultOutputCoder() {
+				return KvCoder.of(NatsIO.DEFAULT_CODER, NatsIO.DEFAULT_CODER);
+			}
+
+			@Override
+			public void validate() {
+				Preconditions.checkNotNull(props.getProperty("servers"), "need to set the 'servers' property");
+				Preconditions.checkNotNull(subject, "need to set the 'subject'");
+			}
+
+			public static class CommitMark implements UnboundedSource.CheckpointMark, Serializable {
+				private static final long serialVersionUID = 2L;
+
+				public void finalizeCheckpoint() throws IOException {}		
+			}
+				
+			private class NatsReader extends UnboundedReader<KV<String,String>> {
+				private String subject;
+				private Integer sid;
+				private ConcurrentLinkedQueue<KV<String,String>> queue;
+				private KV<String,String> current;
+				
+				public NatsReader(String subject) {
+					this.subject = subject;
+					queue = new ConcurrentLinkedQueue<KV<String,String>>();
+				}
+				
+				@Override
+				public boolean advance() throws IOException {
+					// Polling a queue to check whether an update is available in the queue.
+					current = queue.poll();
+					if (current == null)
```

Comment:
totally optional, but Google java style is to always include `{}` with `if` statements or `for` loops, etc. Just helps prevent errors.